File: routing.cpp
#include "routing.h"
#include <algorithm>

RoutingEngine::RoutingEngine(std::shared_ptr<TargetAccessibilityTracker> tracker, RoutingMode mode)
    : tracker_(tracker), mode_(mode) {
}

void RoutingEngine::set_mode(RoutingMode mode) {
    std::lock_guard<std::mutex> lock(mode_mutex_);
    mode_ = mode;
}

RoutingMode RoutingEngine::get_mode() const {
    std::lock_guard<std::mutex> lock(mode_mutex_);
    return mode_;
}

std::shared_ptr<Runway> RoutingEngine::select_runway(
    const std::string& target,
    const std::vector<std::shared_ptr<Runway>>& runways) {
    
    std::lock_guard<std::mutex> lock(mode_mutex_);
    RoutingMode current_mode = mode_;
    
    // Get accessible runways
    std::vector<std::string> accessible_ids = tracker_->get_accessible_runways(target);
    if (accessible_ids.empty()) {
        return nullptr;
    }
    
    // Filter runways to only accessible ones
    std::vector<std::shared_ptr<Runway>> accessible_runways;
    for (const auto& runway : runways) {
        if (std::find(accessible_ids.begin(), accessible_ids.end(), runway->id) != accessible_ids.end()) {
            accessible_runways.push_back(runway);
        }
    }
    
    if (accessible_runways.empty()) {
        return nullptr;
    }
    
    switch (current_mode) {
        case RoutingMode::Latency:
            return select_by_latency(target, accessible_runways);
        case RoutingMode::FirstAccessible:
            return select_first_accessible(target, accessible_runways);
        case RoutingMode::RoundRobin:
            return select_round_robin(target, accessible_runways);
        default:
            return select_first_accessible(target, accessible_runways);
    }
}

std::shared_ptr<Runway> RoutingEngine::select_by_latency(
    const std::string& target,
    const std::vector<std::shared_ptr<Runway>>& runways) {
    
    std::shared_ptr<Runway> best_runway = nullptr;
    double best_latency = 1e9;
    
    for (const auto& runway : runways) {
        auto metrics = tracker_->get_metrics(target, runway->id);
        if (metrics && metrics->avg_response_time > 0.0) {
            if (metrics->avg_response_time < best_latency) {
                best_latency = metrics->avg_response_time;
                best_runway = runway;
            }
        }
    }
    
    if (best_runway) {
        return best_runway;
    }
    
    // Fallback to first accessible
    return select_first_accessible(target, runways);
}

std::shared_ptr<Runway> RoutingEngine::select_first_accessible(
    const std::string& /*target*/,
    const std::vector<std::shared_ptr<Runway>>& runways) {
    
    if (runways.empty()) {
        return nullptr;
    }
    return runways[0];
}

std::shared_ptr<Runway> RoutingEngine::select_round_robin(
    const std::string& target,
    const std::vector<std::shared_ptr<Runway>>& runways) {
    
    if (runways.empty()) {
        return nullptr;
    }
    
    std::lock_guard<std::mutex> lock(rr_mutex_);
    size_t& index = round_robin_index_[target];
    std::shared_ptr<Runway> selected = runways[index % runways.size()];
    index = (index + 1) % runways.size();
    return selected;
}

File: src/routing.rs
use std::collections::HashMap;
use std::sync::Arc;
use parking_lot::RwLock;

use crate::runway::Runway;
use crate::tracker::{TargetAccessibilityTracker, TargetMetrics};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RoutingMode {
    Latency,
    FirstAccessible,
    RoundRobin,
}

pub struct RoutingEngine {
    tracker: Arc<TargetAccessibilityTracker>,
    mode: Arc<RwLock<RoutingMode>>,
    round_robin_index: Arc<RwLock<HashMap<String, usize>>>,
}

impl RoutingEngine {
    pub fn new(tracker: Arc<TargetAccessibilityTracker>, mode: RoutingMode) -> Self {
        Self {
            tracker,
            mode: Arc::new(RwLock::new(mode)),
            round_robin_index: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn set_mode(&self, mode: RoutingMode) {
        *self.mode.write() = mode;
        log::info!("Routing mode changed to {:?}", mode);
    }

    pub fn mode(&self) -> Arc<RwLock<RoutingMode>> {
        Arc::clone(&self.mode)
    }

    pub async fn select_runway(
        &self,
        target: &str,
        runways: &[Runway],
    ) -> Option<Runway> {
        let accessible_ids = self.tracker.get_accessible_runways(target).await;
        
        if accessible_ids.is_empty() {
            return None;
        }

        let accessible_runways: Vec<&Runway> = runways
            .iter()
            .filter(|r| accessible_ids.contains(&r.id))
            .collect();

        if accessible_runways.is_empty() {
            return None;
        }

        let mode = *self.mode.read();
        match mode {
            RoutingMode::Latency => self.select_by_latency(target, &accessible_runways).await,
            RoutingMode::FirstAccessible => Some(accessible_runways[0].clone()),
            RoutingMode::RoundRobin => self.select_round_robin(target, &accessible_runways).await,
        }
    }

    async fn select_by_latency(&self, target: &str, runways: &[&Runway]) -> Option<Runway> {
        let mut best: Option<(&Runway, f64)> = None;

        for runway in runways {
            if let Some(metrics) = self.tracker.get_metrics(target, &runway.id).await {
                if metrics.avg_response_time > 0.0 {
                    let is_better = best
                        .as_ref()
                        .map(|(_, time)| metrics.avg_response_time < *time)
                        .unwrap_or(true);
                    
                    if is_better {
                        best = Some((runway, metrics.avg_response_time));
                    }
                }
            }
        }

        best.map(|(r, _)| r.clone())
            .or_else(|| runways.first().map(|r| (*r).clone()))
    }

    async fn select_round_robin(&self, target: &str, runways: &[&Runway]) -> Option<Runway> {
        if runways.is_empty() {
            return None;
        }

        let mut index_map = self.round_robin_index.write();
        let index = index_map.entry(target.to_string()).or_insert(0);
        let selected = runways[*index % runways.len()].clone();
        *index = (*index + 1) % runways.len();
        Some(selected)
    }
}
